drake/systems/DrakeSystem.cpp

```diff
@@ -0,0 +1,138 @@
+  if (!continuous_state_frame) continuous_state_frame = make_shared<CoordinateFrame>(name+"ContinuousState");
+  if (!discrete_state_frame) discrete_state_frame = make_shared<CoordinateFrame>(name+"DiscreteState");
+  if (!input_frame) input_frame = make_shared<CoordinateFrame>(name+"Input");
+  if (!output_frame) output_frame = make_shared<CoordinateFrame>(name+"Output");
+  state_frame = make_shared<MultiCoordinateFrame>(name+"State",initializer_list<CoordinateFramePtr>({continuous_state_frame,discrete_state_frame}));
+}
+
+DrakeSystem::DrakeSystem(const std::string &_name, unsigned int num_continuous_states, unsigned int num_discrete_states,
+                         unsigned int num_inputs, unsigned int num_outputs)
+        : name(_name),
+          input_frame(nullptr),
+          continuous_state_frame(nullptr),
+          discrete_state_frame(nullptr),
+          output_frame(nullptr) {
+  input_frame = make_shared<CoordinateFrame>(name+"Input",num_inputs,"u");
+  continuous_state_frame = make_shared<CoordinateFrame>(name+"ContinuousState",num_continuous_states,"xc");
+  discrete_state_frame = make_shared<CoordinateFrame>(name+"DiscreteState",num_discrete_states,"xd");
+  output_frame = make_shared<CoordinateFrame>(name+"Output",num_outputs,"y");
+  state_frame = make_shared<MultiCoordinateFrame>(name+"State",initializer_list<CoordinateFramePtr>({continuous_state_frame,discrete_state_frame}));
+}
+
+DrakeSystem::VectorXs DrakeSystem::getRandomState(void) {
+  return DrakeSystem::VectorXs::Random(state_frame->getDim());
+}
+
+DrakeSystem::VectorXs DrakeSystem::getInitialState(void) {
+  return getRandomState();
+}
+
+void DrakeSystem::simulate(double t0, double tf, const DrakeSystem::VectorXs &x0, const SimulationOptions& options) {
+  ode1(t0,tf,x0,options);
+}
+
+#include "timeUtil.h"
+
+inline void handle_realtime_factor(double wall_clock_start_time, double sim_time, double realtime_factor)
+{
+  if (realtime_factor>=0.0) {
+    double time_diff = (getTimeOfDay() - wall_clock_start_time) * realtime_factor - sim_time;
+//    cout << "real time: " << (getTimeOfDay() - wall_clock_start_time) << ", sim time: " << sim_time << endl;
+    if (time_diff <= 0.0) {
+      nanoSleep(-time_diff);
+    } else if (time_diff > 1.0) { // then I'm behind by more than 1 second
+      throw runtime_error(
+              "Not keeping up with requested realtime_factor!  At time " + to_string(sim_time) + ", I'm behind by " +
+              to_string(time_diff) + " sec.");
+    }
+  }
+}
+
+
+void DrakeSystem::ode1(double t0, double tf, const DrakeSystem::VectorXs& x0, const SimulationOptions& options) {
+  double t = t0, dt;
+  double wall_clock_start_time = getTimeOfDay();
+  DrakeSystem::VectorXs x = x0;
+  DrakeSystem::VectorXs u = DrakeSystem::VectorXs::Zero(input_frame->getDim());
+  DrakeSystem::VectorXs y(output_frame->getDim());
+  while (t<tf) {
+    handle_realtime_factor(wall_clock_start_time, t, options.realtime_factor);
```

Comment:
I think we should somehow pull the realtime handling out of the integrator. Otherwise every single type of integrator will have to have this same boilerplate code.